### DISCUSSION TITLE
[@types/karma] should not have log4js as dependency

### DISCUSSION THREAD

**Discussion by scottschafer** [DefinitelyTyped/DefinitelyTyped#58211] [Issues with a @types package]
Or should it?

When I installed this, I thought I was just installing "types". I didn't realize it would import other dependencies into my project.

In our case, this pulled in the following, which stomped on older versions of libraries we were using, causing random run-time errors. 

```
├─┬ @types/karma@6.3.2
│ └─┬ log4js@6.3.0
│   ├── date-format@3.0.0 
│   └─┬ streamroller@2.2.4
│     └── date-format@2.1.0 
```

[a comment by an automated account is not shown]

**Comment by peterblazejewicz** [comment]
Hey, 
Karma depends on this to implement loggers, it's non-optional deps:
https://github.com/karma-runner/karma/blob/master/package.json#L439
Types are just reusing options from `log4js`. v5 of Karma has v4 of `log4js` dependency

**Comment by scottschafer** [comment]
I understand that karma has a dependency on log4js. However, I would not
expect that simply adding @types/karma to my project would change my
project's dependencies aside from just adding type definitions. I could
understand @types/karma having a dependency on @types/log4js (if such a
thing existed) but it seems wrong to me that adding a type definition could
break my run-time behavior by pulling in non-type dependencies.

In my case, I'm adding Typescript to an old and large project that cannot
be easily upgraded. I had added the wrong version of @types/karma, which
had a cascading effect that broke my application at runtime. So this was my
error, and yet it still seems wrong that it happened. Do I need to check
the dependencies on all @types/xx I add to make sure this problem doesn't
happen again?

**Reply by devoto13** [reply]
`karma` exposes some `log4js` type definitions on its config (as it passes the logger configuration through) and `log4js` happen to have type definitions included instead of a separate `@types/log4js` package. As far as I understand, that's how types should work in such a situation. I don't think it is feasible to maintain an up-to-date copy of relevant `log4js` types inside `@types/karma` package and I'm not sure of a better way to do that. Maybe the TypeScript team maintaining DefinitelyTyped has a better suggestion.

Generally, I would not expect this to be a problem: if your project depends on `log4js` X and `@types/karma` depends on `log4js` Y, then the version Y would never end up at the top level `node_modules` and won't be importable by your code. So I'm not quite sure how could it happen unless there was already something wrong with the dependency tree.

> Do I need to check the dependencies on all @types/xx I add to make sure this problem doesn't happen again?

IMO it is always a good idea to check what has been added to the dependency tree when adding new dependencies (not just `@types` packages).

**Comment by scottschafer** [comment]
It's hard for me to know exactly what the problem is. I'm dealing with a
large and complex code base that's up to 12 years old in places.

I eventually tracked down the problem to this. It seems like we have to use
date-format 1.2.0 or tests break. We don't directly depend on date-format
but I guess I can see how this could break things.

I just used "yarn add -D @types/karma". It did occur to me at the time to
specify the version, but I had the thought "it probably doesn't matter
because these are just type definitions." That laziness cost me several
days.

├─┬ @***@***.***
│ └─┬ ***@***.***
│   ├── ***@***.***
│   └─┬ ***@***.***
│     └── ***@***.***
└─┬ ***@***.***
  └─┬ ***@***.***
    ├── ***@***.***
    └─┬ ***@***.***
      └── ***@***.***  deduped
